File: src/main/java/net/maple/handlers/user/UserQuestRequestHandler.java
package net.maple.handlers.user;

import client.Client;
import client.player.quest.QuestRequest;
import net.maple.handlers.PacketHandler;
import scripting.quest.QuestConversationManager;
import scripting.quest.QuestScriptManager;
import util.HexTool;
import util.packet.PacketReader;

public class UserQuestRequestHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        System.out.println("[UserQuestRequestHandler] " + HexTool.toHex(reader.getData()));

        byte action = reader.readByte();
        short questId = reader.readShort();
        int npcId = reader.readInteger();
        short ptUserPosX = reader.readShort();
        short ptUserPosY = reader.readShort();

        if (action == QuestRequest.OPENING_SCRIPT.getValue()) {
            QuestScriptManager.getInstance().converse(c, npcId, questId, true);
        } else if (action == QuestRequest.COMPLETE_SCRIPT.getValue()) {
            QuestScriptManager.getInstance().converse(c, npcId, questId, false);
        } else if (action == QuestRequest.RESIGN_QUEST.getValue()) {
            c.getCharacter().forfeitQuest(questId);
        } else {
            System.out.println("Unknown/unhandled quest action (" + action + ")");
            if (!c.isAdmin()) {
                c.close(this, "Triggered unused quest action qid: " + questId + " - action: " + action);
            }
        }
    }
}

File: src/main/java/net/maple/handlers/user/UserQuestRequestHandler.kt
package net.maple.handlers.user

import client.Client
import client.player.quest.QuestRequest
import net.maple.handlers.PacketHandler
import scripting.dialog.quest.QuestScriptManager
import util.HexTool.toHex
import util.packet.PacketReader

class UserQuestRequestHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        println("[UserQuestRequestHandler] " + toHex(reader.data))

        val action = reader.readByte()
        val questId = reader.readShort()
        val npcId = reader.readInteger()
        val ptUserPosX = reader.readShort()
        val ptUserPosY = reader.readShort()

        when(action.toInt()) {
            QuestRequest.ACCEPT_QUEST.value,
            QuestRequest.OPENING_SCRIPT.value -> openQuest(c, questId.toInt(), npcId)
            QuestRequest.COMPLETE_QUEST.value,
            QuestRequest.COMPLETE_SCRIPT.value -> openQuest(c, questId.toInt(), npcId, false)
            QuestRequest.RESIGN_QUEST.value -> c.character.forfeitQuest(questId.toInt())
            else -> {
                println("Unknown/unhandled quest action ($action)")
                if (!c.isAdmin) {
                    c.close(this, "Triggered unused quest action qid: $questId - action: $action")
                }
            }
        }
    }

    companion object {

        fun openQuest(c: Client, qid: Int, npc: Int, start: Boolean = true) {
            c.script = null
            QuestScriptManager[qid]?.let {
                if (start) {
                    it.execute(c)
                } else {
                    it.finish(c)
                }
            } ?: run {
                c.character.writeNpc(
                    npc,
                    "This quest does not appear to have a script\r\n" +
                            "Please report this to a staff member\r\n" +
                            "QID: #r" + qid + "#k\r\n" +
                            "NPC: #r" + npc + "#k\r\n" +
                            "Map: #r" + c.character.fieldId
                )
            }
        }
    }
}